### DISCUSSION TITLE
_cfv  - help

### DISCUSSION THREAD

**Discussion by Deanobeano2** [cortexrd/Knack-Toolkit-Library#404] [Q&A] [answered]
Hi

I'm wanting to highlight a menu button if a criteria is met. In this case I'd like to highlight the "Archived Notes" button in some way if the field "Has Archived Notes" is Yes.

![image](https://github.com/user-attachments/assets/24814d52-b227-4a94-9b64-a7054f0e1ddd)

How do I target the correct menu button for the _cfv=[.....] statement? I believe the conditional statement will be [ktlCond, is, Yes, Has Archived Notes]. Once this is working then I'd hide the "Has Archived Notes" field.

Dean

**Comment by CSWinnall** [comment]
Hi Dean

Try this

'_cfv=[Has Archived Notes, is, Yes, fontColor, backgroundColor], [ktlTarget, $('#view_123 li.contains("Archived Notes")')]`

Replace fontColor & backgroundColor with your colors & 123 with the view id of the menu view.

Let us know how it goes.
Craig

**Comment by Deanobeano2** [comment]
Hi Craig

Thank you for offering a solution.

I tried the following unsuccessfully:
'_cfv=[Has Archived Notes, is, Yes, black, red], [ktlTarget, $('#view_347 li.contains("Archived Notes")')]'

'_cfv=[Has Archived Notes, is, Yes, black, red, bold,], [ktlTarget, $('#view_347 li.contains("Archived Notes")')]'

I broke it down to this '_cfv=[Has Archived Notes, is, Yes, black, red, bold,]' which shows the _cfv is working so the issue is targeting the menu button.

![image](https://github.com/user-attachments/assets/389db8bb-ab95-4417-b538-f218c4dc356f)

Do you have any other thoughts?

Dean

**Comment by CSWinnall** [comment]
Try a:contains some versions of knack are different
Craig

**Comment by Deanobeano2** [comment]
Hi Craig

The braincells started working and I finally found where Norm had helped me with a similar issue. I removed the archive icon and used this which resulted in the desired result, for the meantime:

'_cfv=[Has Archived Notes, is, Yes, black, red, bold,], [ktlTarget, $('#view_347 .knMenuLink:textEquals("Archived Notes")')] 

![image](https://github.com/user-attachments/assets/af5d65be-dad3-4913-9f6c-a6bd4c9020ed)

Do you know if it's possible to add the archive icon back in and still achieve this result?
"<span class="knMenuLink__icon knMenuLink__icon--isLeft"><i class="fa fa-archive"></i></span>"

Dean

**Comment by cortexrd** [comment]
Hi @Deanobeano2,

I was just about to ask you to try the textEquals syntax.
Now why it wipes out your icon?... good question.  It's certainly a KTL bug.
Can you share me that page?  I still have your app link and login.

ND

**Comment by cortexrd** [comment]
It's ok, I found that page.

**Comment by cortexrd** [comment]
Dean,

It seems that there is no icon in the first place!  Even when I bypass the KTL, there's nothing.  See it here:

![image](https://github.com/user-attachments/assets/b8bd0e70-1222-4c12-83f1-d5a73a0f0aaa)

Norm

**Comment by Deanobeano2** [comment]
Hi Norm

Sorry, you're too fast for me. There's no bug. I removed the icon to eliminate variables and make it easier to copy the code you'd supplied previously.

What I'd now like to do, if possible, is add the icon back but assume that will mean changing the criteria from "textEquals" to something else.

What would that be?

Dean

**Comment by cortexrd** [comment]
I see what you mean.

1. When the icon is there the textEquals doesn't find it
2. But it should ignore the icon and just compare the text

There's a bug and I'll fix it.

**Comment by Deanobeano2** [comment]
Hi Norm

It stops working when the Archive icon is re-introduced, sorry.

Dean

**Comment by cortexrd** [comment]
I got it.  Try Dev version in 10 seconds.

**Comment by cortexrd** [comment]
Can you put back the icon?

**Comment by Deanobeano2** [comment]
The icon is back.

To run the Dev version do I select KTL Code and then enter Dev?

**Comment by Deanobeano2** [comment]
If so, I see a flash of, the now, orange background before it finally renders with none of the desired formatting.

**Comment by cortexrd** [comment]
@Dean,

This is becoming crazy and hard to debug!!!!

Sometimes you add the icon with the text (top arrow), then the next line you add the icon in the menu (bottom arrow).

![image](https://github.com/user-attachments/assets/42b0e5b7-d62e-48e9-8ccc-2ac2dfea0406)

So, the reason why it didn't work in the first place is because the icon is part of the text.  If you add the icon with the text, the code will work properly, but only if you add the same icon in your comparison string in the _cfv keyword.  Both must match.

Otherwise, add the icon with the menu button and it will also work.

On my side, I'll leave the code as it was.

**Comment by cortexrd** [comment]
Go back to Prod version, forget about Dev for now.

**Comment by Deanobeano2** [comment, marked as answer]
Sorry, I've always just used the Knack standard default for adding an icon.

![image](https://github.com/user-attachments/assets/973da5d5-b3ca-4e2d-8ac3-d5e564f84037)

**Reply by CSWinnall** [reply]
Hi Dean

We do this a lot and contains should work fine with menu links with icons.

```
<br />_cls=[ktlDisplayNone], [ktlCond, has, Admis Data, Archived Data, Hidden Client Details], [ktlTarget, $('#view_7539 .menu-links__list-item:contains("Queries"), .view-group:has(#view_7952), .menu-links__list-item:contains("Edit Pre-Assessment Questions"), #view_7861')] _notes=if archived hide complete assess, parts of assess & queries

<br /> _cls=[ktlDisplayNone], [ktlCond, has, False, New Admission, Hidden Client Details], [ktlTarget, $('#view_7539 .menu-links__list-item:contains("Queries")')] _notes = hide queries if old admis

<br />_cfv=[field_3940, lt, 2,,,,r], [ktlTarget, $('#view_7539 .menu-links__list-item:contains("Queries"), #view8761')]
_notes=hide queries & view assess if assess part < 2

<br />_cfv=[Num Episodes, lt, 2,,,,r], [ktlTarget, $('#view_7539 .menu-links__list-item:contains("Previous Episodes")')]
```

To get around the issue of the icon we use either li:contains, .menu-links__list-item:contains.  Text Equals will not work here brcause of the icon.  With contains the text must be an exact match. but you only need to match some of it.
Craig

**Reply by Deanobeano2** [reply]
Hi Craig

Thank you for all of the examples.

I believe I actually tried textContains thinking it would a logical follow-on to textEquals but when it didn't work thought maybe I had the syntax wrong. All the while it was some Javascript I'd forgotten about, when fooling around with this issue, interfering with KTL code. Another rookie mistake to learn from.

Dean

**Comment by cortexrd** [comment]
Ok, but about the other one to the left:  Jobs.
Can you show me the icon and the text?

I suspect that the icon is part of the text like this: "🧹Jobs", and not a separate icon like you've shown here.

**Comment by Deanobeano2** [comment]
Sorry, it's all working now. I'd left some in some Javascript that was interfering with KTL.

**Comment by cortexrd** [comment]
Ok then, maybe this was part of the issue.

But still, you must remember that if you include some icon in your label, you must also include that same icon in the condition for a keyword.

If all good, mark this post as the Anwser so it closes please.

Cheers,
Norm

**Comment by Deanobeano2** [comment]
![image](https://github.com/user-attachments/assets/a2b93ac8-8f98-4ec9-9f16-f15854f55a26)

How would you specify the icon in the condition for the keyword in this case?

**Comment by cortexrd** [comment]
@Deanobeano2,

It would be like this:

_cfv=[Has Archived Notes, is, Yes, black, #ffbb73, bold,], [ktlTarget, $('#view_347 .knMenuLink:textEquals("📦Archived Notes")')]

Assuming the label of your button would also be "📦Archived Notes".

Then, all your icons (emojis in this case) would have the same style, with colors.

As Craig pointed out, you can also use **:contains** instead of **:textEquals**.  This would be more permissive and find the text successfully because it doesn't require to be an exact match, ignoring the icon.  But then it might select the wrong element if you had another one like "📦Un-Archived Notes", for example.

**Reply by Deanobeano2** [reply]
Hi Norm

This is the working code even though the icon is in the menu button.

"_cfv=[Has Archived Notes, is, Yes, black, #ffbb73, bold,], [ktlTarget, $('#view_347 .knMenuLink:textEquals("Archived Notes")')]"

By some chance does it ignore the icon as it's not text?

Dean

**Reply by cortexrd** [reply]
Yes, this what I was trying to explain yesterday.

Your current code works as expected because the icon is part of the button, but **not within the text itself**.

You can see this when you press Ctrl+Shift+C and hover over the icons for these two buttons.

1. The icon is separate from the text

![image](https://github.com/user-attachments/assets/bb32300a-3088-4830-a13f-9ee674bf8e9f)


2. The icon is part of the text

![image](https://github.com/user-attachments/assets/9c7c4ce4-7a0e-4071-bfb0-7aefeac8db1a)

Norm

**Comment by Deanobeano2** [comment]
Hi Norm

Thank you for the tip on how to inspect the elements as I never knew you could do it that way.

Dean
